File: lambda/dpm_splunk_logger_py/splunk.py
import os
import time
import boto3
import json
import urllib.request
from urllib.parse import urlparse
from botocore.vendored import requests
from urllib.error import HTTPError, URLError
from urllib.request import urlopen, Request
from urllib.parse import urlencode


# Define logger config using environment variables
logger_config = {
    'url': "https://http-inputs-br.splunkcloud.com/services/collector",
    'token': "56a47b4b-12e2-4855-a687-2b8224a39ef0"
}

logger_config_dev = {
    'url': 'https://http-inputs-br-nprd.splunkcloud.com/services/collector',
    'token': '47ae2988-d442-1344-498a-d18179ff4adb'
}

os.environ["application_name"] = 'Data Preparation Manager'     
os.environ["application_short_name"] = 'DPM'
message_data = ""
os_environ_keys=[]

def store_message(event):
    message_data = message_data + "\n" + event
    
def flush(event,context):
    json_data = json.loads(event)
    json_data["logs"] = message_data
    log_message(json_data,context)

def log_message(event, context):
    print('Received event:', event)
    extract_event_setenviron(event, context) 
    data ={}

    # Log JSON objects to Splunk
    # logger.log(event)

    # Log JSON objects with optional 'context' argument (recommended)
    # This adds valuable Lambda metadata including functionName as source, awsRequestId as field
    # logger.log(event, context)

    # Log strings
    # logger.log('value1 = {}'.format(event['key1']), context)

    # Log with user-specified timestamp - useful for forwarding events with embedded
    # timestamps, such as from AWS IoT, AWS Kinesis, AWS CloudWatch Logs
    # Change "int(time.time() * 1000)" below to event timestamp if specified in event payload
    # logger.log_with_time(int(time.time() * 1000), event, context)

    # Advanced:
    # Log event with user-specified request parameters - useful to set input settings per event vs token-level
    # Full list of request parameters available here:
    # http://docs.splunk.com/Documentation/Splunk/latest/RESTREF/RESTinput#services.2Fcollector
    
   
    my_session = boto3.session.Session()
    my_region = my_session.region_name
    os.environ["aws_account_id"] = context.split(':')[4]
    os.environ["aws_region"] = my_region
    
    os_environ_keys.append("application_name")
    os_environ_keys.append("application_short_name")
    os_environ_keys.append("aws_account_id")
    os_environ_keys.append("aws_region")
    
    
    account_id_value = event["aws_account_id"]
    current_stage = 'PROD'
    
    if account_id_value == "187777304606" :
        current_stage = current_stage.replace('PROD','DEV')    
    elif account_id_value == "471112732183" :
        current_stage = current_stage.replace('PROD','PROD')    
    elif account_id_value == "556144470667" :
        current_stage = current_stage.replace('PROD','QA')
        
    if context.function_name.find("dpmdev-di") != -1:
        current_stage = current_stage.replace('DEV','INT')
    elif context.function_name.find("br_icsuat") != -1:
        current_stage = current_stage.replace('QA','UAT')
    
    
    os.environ["stage"]= current_stage
    os_environ_keys.append("stage")
    
    if current_stage == 'DEV' or current_stage == 'INT':
       logger = Logger(logger_config_dev)
    else:
       logger = Logger(logger_config)
    
    for key in os_environ_keys:
        if key in os.environ:
            data[key] = os.environ[key]
    
    logger.log_event({
        'time': int(time.time() * 1000),
        'host': 'serverless',
        'source': 'lambda:{}'.format(context.function_name),
        'sourcetype': 'httpevent',
        'event': data
    })

    # Send all the events in a single batch to Splunk
    logger.flush_async()

class Logger:
    def __init__(self, config):
        self.url = config['url']
        self.token = config['token']

        self.addMetadata = True
        self.setSource = True

        self.parsedUrl = urlparse(self.url)

        self.requestOptions = {"Content-Type": "application/json",
		'Authorization': 'Splunk ' + self.token
		}

        self.payloads = ""

    def log_event(self, payload):
        print('Request payload: ' + json.dumps(payload))
        self.payloads=json.dumps(payload)

    def flush_async(self, callback=None):
        callback = callback or (lambda: None)

        print('Sending event(s)')

        data = (self.payloads)
        data = data.encode('utf-8')
        req = urllib.request.urlopen(urllib.request.Request(''+self.url,data,self.requestOptions))
        res = req.getcode()
        print('Response code is: ' + str(res))

        if res != 200:
            error = ValueError('Error: status code = %d\n\n%s' % (res, req.read()))
            print(error)
        else:
            self.payloads = []
def extract_event_setenviron(event_data, context):    
    try:
        for key, value in event_data.items():
            os.environ[key] = value 
            os_environ_keys.append(key)       
    except Exception as e:
        raise ValueError("Failed to extract event info: " + str(e))

File: src/main/java/com/broadridge/ebcdic/Splunk.java

package com.broadridge.ebcdic;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URI;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;
import com.fasterxml.jackson.databind.ObjectMapper;

public class Splunk {

	private final String url;
	private final String token;
	private String payload;

	public Splunk(String url, String token) {
		this.url = url;
		this.token = token;
		this.payload = "";
	}

	public void logEvent(Map<String, Object> data) {
		data.put("time", Instant.now().getEpochSecond());
		//System.out.println(Instant.now().getEpochSecond());
		//System.out.println(Instant.now());
		data.put("host", "serverless");
		data.put("source", "dpm-di-docker:Ebcdic to Json");
		data.put("sourcetype", "httpevent");
		this.payload = mapToJson(data);
		//System.out.println(this.payload);
	}

	private String mapToJson(Map<String, Object> data) {
		ObjectMapper objectMapper = new ObjectMapper();
		Map<String, Object> JsonData = new HashMap<>();
		JsonData.put("time", Instant.now().getEpochSecond());
		//System.out.println(Instant.now().getEpochSecond());
		//System.out.println(Instant.now());
		JsonData.put("host", "serverless");
		JsonData.put("source", "dpm-di-docker:Ebcdic to Json");
		JsonData.put("sourcetype", "httpevent");
		JsonData.put("event", data);
		try {
			return objectMapper.writeValueAsString(JsonData);
		} catch (Exception e) {
			e.printStackTrace();
		}
		return null;
	}

	public void flushAsync(Runnable callback) {
		Runnable effectiveCallback = (callback != null) ? callback : () -> {
		};

		new Thread(() -> {
			try {
				URI uri = new URI(url);
				HttpURLConnection connection = (HttpURLConnection) uri.toURL().openConnection();
				connection.setRequestMethod("POST");
				connection.setRequestProperty("Content-Type", "application/json");
				connection.setRequestProperty("Authorization", "Splunk " + token);
				connection.setDoOutput(true);

				try (OutputStream os = connection.getOutputStream()) {
					// os.write(payload.getBytes(StandardCharsets.UTF_8));
					byte[] input = payload.getBytes("utf-8");
					os.write(input, 0, input.length);
				}

				int statusCode = connection.getResponseCode();
				//System.out.println("Response code is: " + statusCode);

				try (BufferedReader reader = new BufferedReader(
						new InputStreamReader(getInputStreamOrErrorStream(connection, statusCode)))) {
					String response = reader.lines().reduce("", (acc, line) -> acc + line);
					//System.out.println("Response body: " + response);
				}

				effectiveCallback.run();
			} catch (Exception e) {
				e.printStackTrace();
			}
		}).start();
	}

	private InputStream getInputStreamOrErrorStream(HttpURLConnection connection, int statusCode) throws IOException {
		return (statusCode >= 200 && statusCode < 400) ? connection.getInputStream() : connection.getErrorStream();
	}

	public static void logMessage(Map<String, Object> input, String accountData) {

		Map<String, Object> data = new HashMap<>(input);
		data.put("application_name", "Data Preparation Manager");
		data.put("application_short_name", "DPM");
		
		data.put("aws_region", accountData.trim().split(" ")[1]);
	
		String currentStage = determineStage((String) accountData.trim().split(" ")[0]);
		//System.out.println("==============>"+ accountData.trim().split(" ")[0] + " account " +accountData);
		data.put("stage", currentStage);	
		//System.out.println("stage ==>" + currentStage);

		Splunk customLogger = new Splunk(getSplunkUrl(currentStage), getSplunkToken(currentStage));

		customLogger.logEvent(data);
		customLogger.flushAsync(() -> System.out.println("Async operation completed"));
	}

	public static String determineStage(String awsAccountId) {
		if ("187777304606".equals(awsAccountId))
			return "DEV";
		if ("471112732183".equals(awsAccountId))
			return "PROD";
		if ("556144470667".equals(awsAccountId))
			return "QA";
//		if (functionName.contains("dpmdev-di"))
//			return "INT";
//		if (functionName.contains("br_icsuat"))
//			return "UAT";
		return "PROD";
	}

	private static String getSplunkUrl(String stage) {
		return (stage.equals("DEV") || stage.equals("INT"))
				? "https://http-inputs-br-nprd.splunkcloud.com/services/collector"
				: "https://http-inputs-br.splunkcloud.com/services/collector";
	}

	private static String getSplunkToken(String stage) {
		return (stage.equals("DEV") || stage.equals("INT")) ? "47ae2988-d442-1344-498a-d18179ff4adb"
				: "56a47b4b-12e2-4855-a687-2b8224a39ef0";
	}

	
}